### PR TITLE
Move Handshake to Kotlin

### PR DESCRIPTION
**Pull request by shaishavgandhi05**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+ * is its *peer*.
+ *
+ * This value object describes a completed handshake. Use [ConnectionSpec] to set policy
+ * for new handshakes.
+ */
+class Handshake private constructor(
+    private val tlsVersion: TlsVersion,
+    private val cipherSuite: CipherSuite,
+    private val peerCertificates: List<Certificate>,
+    private val localCertificates: List<Certificate>
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: nit: these should be +2 not +4

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  fun cipherSuite() = cipherSuite
+
+  /** Returns a possibly-empty list of certificates that identify the remote peer. */
+  fun peerCertificates() = peerCertificates
+
+  /** Returns the remote peer's principle, or null if that peer is anonymous. */
+  fun peerPrincipal(): Principal? {
+    return if (peerCertificates.isNotEmpty())
+      (peerCertificates[0] as X509Certificate).subjectX500Principal
+    else
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: curlies on both branches

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /** Returns a possibly-empty list of certificates that identify this peer. */
+  fun localCertificates() = localCertificates
+
+  /** Returns the local principle, or null if this peer is anonymous. */
+  fun localPrincipal(): Principal? {
+    return if (localCertificates.isNotEmpty())
+      (localCertificates[0] as X509Certificate).subjectX500Principal
+    else
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: curlies

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      null
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (other !is Handshake) return false
+    return (tlsVersion == other.tlsVersion
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: redundant parens. you can also fold `other is Handshake` into `&&` chain

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    return ("Handshake{"
+        + "tlsVersion="
+        + tlsVersion
+        + " cipherSuite="
+        + cipherSuite
+        + " peerCertificates="
+        + names(peerCertificates)
+        + " localCertificates="
+        + names(localCertificates)
+        + '}'.toString())
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: redundant parens, you shouldn't need `.toString()` call either

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+      val peerCertificates: Array<Certificate>? = try {
+        session.peerCertificates
+      } catch (ignored: SSLPeerUnverifiedException) {
+        null
+      }
+
+      val peerCertificatesList = if (peerCertificates != null)
+        Util.immutableList(*peerCertificates)
+      else
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: curlies

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        emptyList()
+
+      val localCertificates = session.localCertificates
+      val localCertificatesList = if (localCertificates != null)
+        Util.immutableList(*localCertificates)
+      else
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: curlies

File: okhttp/src/main/java/okhttp3/Handshake.kt [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    val strings = ArrayList<String>()
+
+    for (cert in certificates) {
+      if (cert is X509Certificate) {
+        strings.add(cert.subjectDN.toString())
+      } else {
+        strings.add(cert.type)
+      }
+    }
+    return strings
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Handshake.kt]: Implementation could be
```kotlin
return certificates
    .map {
      if (it is ..) {
        it.subjectDN.toString()
      } else {
        it.type
      }
    }
```
which will also pre-size the new list

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 1 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: Done 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 2 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: Done 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 3 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: Done 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 4 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: Done 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 5 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 7 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: 26a75752e3edd3600fbc2f4c0e796134e3b08277

File: okhttp/src/main/java/okhttp3/Handshake.kt (hunk 8 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Handshake.kt]: Nice! 26a75752e3edd3600fbc2f4c0e796134e3b08277

**Pull request merged**